### PR TITLE
Add openshift e2e testing to pipeline enabled through linchpin libvirt container

### PR DESCRIPTION
**Pull request by samvarankashyap**
$Subject

### CODE REVIEW

File: config/s2i/create-containers.sh [hunk 1]
```diff
@@ -24,4 +25,6 @@ oc new-app rsync-builder ${REPO_URL_PARAM} ${REPO_REF_PARAM}
 oc new-app ostree-compose-builder ${REPO_URL_PARAM} ${REPO_REF_PARAM}
 oc new-app ostree-image-compose-builder ${REPO_URL_PARAM} ${REPO_REF_PARAM}
 oc new-app singlehost-test-builder ${REPO_URL_PARAM} ${REPO_REF_PARAM}
+oc new-app package-test-builder ${REPO_URL_PARAM} ${REPO_REF_PARAM}
```
Comment (johnbieren) [inline config/s2i/create-containers.sh]: Probably a rebase error, but don't want this line as package-test is gone in favor of singlehost-test

Comment (samvarankashyap) [reply config/s2i/create-containers.sh]: I will change it . However, this shell script should not effect the pipeline job.

Comment (johnbieren) [reply config/s2i/create-containers.sh]: Agreed, it should not. My only point was I think it was accidentally added in and we don't need it as it wouldn't work if the script was run

**Comment by samvarankashyap** [comment]
[test]

**Comment by samvarankashyap** [comment]
[test]

**Comment by samvarankashyap** [comment]
[test]

**Comment by samvarankashyap** [comment]
[test]

**Comment by samvarankashyap** [comment]
[test]

**Comment by samvarankashyap** [comment]
[test]

**Comment by samvarankashyap** [comment]
[test]

**Comment by samvarankashyap** [comment]
[test]

**Comment by p3ck** [comment]
[test] - anaconda install failed. not sure why

**Comment by samvarankashyap** [comment]
Tests ran successfully on f27 build.
but f26 build fails in openshift 3.6 installation . 
the following error is observed while installing 3.6 openshift on f26 .
```
<192.168.125.87> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o 'IdentityFile="/home/.ssh/ex"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publicke
y -o PasswordAuthentication=no -o User=admin -o ConnectTimeout=10 -o ControlPath=/home/.ansible/cp/cdc3e24cb6 -tt 192.168.125.87 '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-qpfrqgoydyv
dykyhsymagkvbscbvvfqa; /usr/bin/python3 /home/admin/.ansible/tmp/ansible-tmp-1509045363.8-147437861082049/systemd.py; rm -rf "/home/admin/.ansible/tmp/ansible-tmp-1509045363.8-147437861082049/" > /dev/null 2>&1'"'"'"'"'"'"'"'"'
 && sleep 0'"'"''                                                                                                                                                                                                                  
<192.168.125.87> (0, '\r\n{"msg": "Unable to restart service origin-master: Job for origin-master.service failed because the control process exited with error code.\\nSee \\"systemctl  status origin-master.service\\" and \\"jou
rnalctl  -xe\\" for details.\\n", "failed": true, "invocation": {"module_args": {"name": "origin-master", "state": "restarted", "daemon_reload": false, "user": false, "no_block": false, "enabled": null, "masked": null}}}\r\n', 
'OpenSSH_7.4p1, OpenSSL 1.0.2k-fips  26 Jan 2017\r\ndebug1: Reading configuration data /root/.ssh/config\r\ndebug1: /root/.ssh/config line 1: Applying options for *\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\nd
ebug1: /etc/ssh/ssh_config line 58: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: re
quest forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 8011\r\ndebug3: mux_client_request_session: sess
ion request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\nShared connection to 192.168.125.8
7 closed.\r\n')                                                                                                                                                                                                                    
fatal: [192.168.125.87]: FAILED! => {                                                                                                                                                                                              
    "changed": false,                                                                                                                                                                                                              
    "failed": true,                                                                                                                                                                                                                
    "invocation": {                                                                                                                                                                                                                
        "module_args": {                                                                                                                                                                                                           
            "daemon_reload": false,                                                                                                                                                                                                
            "enabled": null,                                                                                                                                                                                                       
            "masked": null,                                                                                                                                                                                                        
            "name": "origin-master",                                                                                                                                                                                               
            "no_block": false,                                                                                                                                                                                                     
            "state": "restarted",                                                                                                                                                                                                  
            "user": false                                                                                                                                                                                                          
        }                                                                                                                                                                                                                          
    },                                                                                                                                                                                                                             
    "msg": "Unable to restart service origin-master: Job for origin-master.service failed because the control process exited with error code.\nSee \"systemctl  status origin-master.service\" and \"journalctl  -xe\" for details.
\n"                                                                                                                                                                                                                                
}                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                   
RUNNING HANDLER [openshift_master : restart master api] ***************************************************************************************************************************************************************************
                                                                                                                                                                                                                                   
RUNNING HANDLER [openshift_master : restart master controllers] *******************************************************************************************************************************************************************
META: ran handlers                                                                                                                                                                                                                 
        to retry, use: --limit @/root/linchpin_workspace/hooks/ansible/openshift-ansible/playbooks/byo/config.retry                                                                                                                
                                                                                                                                                                                                                                   
PLAY RECAP ************************************************************************************************************************************************************************************************************************
192.168.125.87             : ok=298  changed=8    unreachable=0    failed=1                                                                                                                                                        
localhost                  : ok=8    changed=0    unreachable=0    failed=0                                                                                                                                                        
                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                   
Failure summary:                                                                                                                                                                                                                   
                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                   
  1. Hosts:    192.168.125.87                                                                                                                                                                                                      
     Play:     Configure masters                                                                                                                                                                                                   
     Task:     restart master                                                                                                                                                                                                      
     Message:  Unable to restart service origin-master: Job for origin-master.service failed because the control process exited with error code.                                                                                   
               See "systemctl  status origin-master.service" and "journalctl  -xe" for details.  
```

ansible-inventory generated : 
```
# Create an OSEv3 group that contains the masters and nodes groups                                                                                                                                                                 
[OSEv3:children]                                                                                                                                                                                                                   
masters                                                                                                                                                                                                                            
nodes                                                                                                                                                                                                                              
                                                                                                                                                                                                                                   
# Set variables common for all OSEv3 hosts                                                                                                                                                                                         
[OSEv3:vars]                                                                                                                                                                                                                       
ansible_user=admin                                                                                                                                                                                                                 
ansible_become=true                                                                                                                                                                                                                
                                                                                                                                                                                                                                   
openshift_master_default_subdomain=192.168.125.87.xip.io                                                                                                                                                                           
                                                                                                                                                                                                                                   
deployment_type=origin                                                                                                                                                                                                             
openshift_deployment_type=origin                                                                                                                                                                                                   
openshift_release=v3.6                                                                                                                                                                                                             
openshift_image_tag=v3.6.0                                                                                                                                                                                                         
#deployment_subtype=registry                                                                                                                                                                                                       
containerized=true                                                                                                                                                                                                                 
ansible_ssh_private_key_file=/home/.ssh/ex                                                                                                                                                                                         
# enable htpasswd auth                                                                                                                                                                                                             
openshift_master_identity_providers=[{'name': 'htpasswd_auth', 'login': 'true', 'challenge': 'true', 'kind': 'HTPasswdPasswordIdentityProvider', 'filename': '/etc/origin/master/htpasswd'}]                                       
openshift_master_htpasswd_users={'admin': '$apr1$zgSjCrLt$1KSuj66CggeWSv.D.BXOA1'}                                                                                                                                                 
                                                                                                                                                                                                                                   
# host group for masters                                                                                                                                                                                                           
[masters]                                                                                                                                                                                                                          
192.168.125.87                                                                                                                                                                                                                     
                                                                                                                                                                                                                                   
# host group for worker nodes, we list master node here so that                                                                                                                                                                    
# openshift-sdn gets installed. We mark the master node as not                                                                                                                                                                     
# schedulable.                                                                                                                                                                                                                     
[nodes]                                                                                                                                                                                                                            
192.168.125.87 openshift_node_labels="{'region': 'infra', 'zone': 'default'}" openshift_schedulable=true    
```

```
[admin@atomic01 ~]$ systemctl status origin-master.service                                                                                                                                                                         
● origin-master.service                                                                                                                                                                                                            
   Loaded: loaded (/etc/systemd/system/origin-master.service; enabled; vendor preset: disabled)                                                                                                                                    
   Active: activating (start-post) (Result: exit-code) since Thu 2017-10-26 19:21:39 UTC; 466ms ago                                                                                                                                
  Process: 8378 ExecStop=/usr/bin/docker stop origin-master (code=exited, status=1/FAILURE)                                                                                                                                        
  Process: 8392 ExecStart=/usr/bin/docker run --rm --privileged --net=host --name origin-master --env-file=/etc/sysconfig/origin-master -v /var/lib/origin:/var/lib/origin -v /var/log:/var/log -v /var/run/docker.sock:/var/run/do
  Process: 8388 ExecStartPre=/usr/bin/docker rm -f origin-master (code=exited, status=1/FAILURE)                                                                                                                                   
 Main PID: 8392 (code=exited, status=125); Control PID: 8393 (sleep)                                                                                                                                                               
    Tasks: 1 (limit: 4915)                                                                                                                                                                                                         
   Memory: 328.0K                                                                                                                                                                                                                  
      CPU: 28ms                                                                                                                                                                                                                    
   CGroup: /system.slice/origin-master.service                                                                                                                                                                                     
           └─control                                                                                                                                                                                                               
             └─8393 /usr/bin/sleep 10                                                                                                                                                                                              
                                                                                                                                                                                                                                   
Oct 26 19:21:39 atomic01.example.com systemd[1]: origin-master.service: Service hold-off time over, scheduling restart.                                                                                                            
Oct 26 19:21:39 atomic01.example.com systemd[1]: Stopped origin-master.service.                                                                                                                                                    
Oct 26 19:21:39 atomic01.example.com systemd[1]: Starting origin-master.service...                                                                                                                                                 
Oct 26 19:21:39 atomic01.example.com docker[8388]: Error response from daemon: No such container: origin-master                                                                                                                    
Oct 26 19:21:39 atomic01.example.com docker[8392]: /usr/bin/docker-current: Error response from daemon: failed to create default sandbox: namespace creation reexec command failed: fork/exec /proc/self/exe: cannot allocate memor
Oct 26 19:21:39 atomic01.example.com systemd[1]: origin-master.service: Main process exited, code=exited, status=125/n/a 
```

**Comment by samvarankashyap** [comment]
Related [#5076](https://github.com/openshift/openshift-ansible/issues/5076)
Fix suggested is to restart origin-master and restart installation gain
```
systemctl restart atomic-openshift-master
```

**Comment by samvarankashyap** [comment]
At last fixed it 💯 !

**Comment by samvarankashyap** [comment]
@scoheb : could you please review it !!

File: Jenkinsfile [hunk 2]
```diff
@@ -262,7 +270,7 @@ podTemplate(name: podName,
                         // Send message org.centos.prod.ci.pipeline.compose.complete on fedmsg
                         pipelineUtils.sendMessageWithAudit(messageFields['properties'], messageFields['content'], msgAuditFile, fedmsgRetryCount)
 
-                        pipelineUtils.checkLastImage(currentStage)
+				pipelineUtils.checkLastImage(currentStage)
```
Comment (scoheb) [inline Jenkinsfile]: nit: whitespace

File: src/org/centos/pipeline/PipelineUtils.groovy [hunk 3]
```diff
@@ -652,6 +652,35 @@ def executeInContainer(String stageName, String containerName, String script) {
...
+ */
+def executeInContainerNoPrep(String stageName, String containerName, String script) {
+    //
+    // Kubernetes plugin does not let containers inherit
+    // env vars from host. We force them in.
+    //
+    containerEnv = env.getEnvironment().collect { key, value -> return key+'='+value }
+    withEnv(containerEnv) {
+        container(containerName) {
+            sh 'pwd'
```
Comment (scoheb) [inline src/org/centos/pipeline/PipelineUtils.groovy]: not needed anymore

File: src/org/centos/pipeline/PipelineUtils.groovy [hunk 4]
```diff
@@ -652,6 +652,35 @@ def executeInContainer(String stageName, String containerName, String script) {
...
+            sh 'ls -l /tmp'
```
Comment (scoheb) [inline src/org/centos/pipeline/PipelineUtils.groovy]: not needed anymore

File: src/org/centos/pipeline/PipelineUtils.groovy [hunk 5]
```diff
@@ -652,6 +652,35 @@ def executeInContainer(String stageName, String containerName, String script) {
...
+            //sh "cp -fv ${WORKSPACE}/fedora.keytab /home/fedora.keytab"
```
Comment (scoheb) [inline src/org/centos/pipeline/PipelineUtils.groovy]: remove

File: src/org/centos/pipeline/PipelineUtils.groovy [hunk 6]
```diff
@@ -652,6 +652,35 @@ def executeInContainer(String stageName, String containerName, String script) {
...
+            sh 'env'
```
Comment (scoheb) [inline src/org/centos/pipeline/PipelineUtils.groovy]: not needed anymore

File: src/org/centos/pipeline/PipelineUtils.groovy [hunk 7]
```diff
@@ -652,6 +652,35 @@ def executeInContainer(String stageName, String containerName, String script) {
...
+            sh script
+            //sh "ls -lR logs || true"
```
Comment (scoheb) [inline src/org/centos/pipeline/PipelineUtils.groovy]: remove

File: src/org/centos/pipeline/PipelineUtils.groovy [hunk 8]
```diff
@@ -652,6 +652,35 @@ def executeInContainer(String stageName, String containerName, String script) {
...
+        }
+    }
+    sh "mkdir -p " + stageName+"/logs"
```
Comment (scoheb) [inline src/org/centos/pipeline/PipelineUtils.groovy]: see recent changes in executeInContainer. Maybe you can use that one now.

File: vars/pipelineUtils.groovy [hunk 9]
```diff
@@ -145,6 +145,10 @@ class pipelineUtils implements Serializable {
     def executeInContainer(stageName, containerName, script) {
         pipelineUtils.executeInContainer(stageName, containerName, script)
     }
+    
+    def executeInContainerNoPrep(stageName, containerName, script) {
```
Comment (scoheb) [inline vars/pipelineUtils.groovy]: see recent changes in executeInContainer. Maybe you can use that one now.

**Comment by samvarankashyap** [comment]
[test]

**Comment by samvarankashyap** [comment]
[test]

**Comment by samvarankashyap** [comment]
[test]

**Comment by scoheb** [comment]
[merge]

**Comment by lioramilbaum** [comment]
I am implementing something similar to the linchpin-libvirt container. Can you please provide some additional information on how this container acts in run time? Does it run on CentOS CI OCP? Does it needs elevated permissions to build the image and run the container? @ifireball

**Pull request merged**